patient_data_source.py

```diff
@@ -0,0 +1,214 @@
+from abc import ABC, abstractmethod
+import names
+import warnings
+from datetime import datetime
+from enum import Enum
+from collections.abc import Iterable
+from fhirclient.models.patient import Patient as FHIR_Patient
+from fhirclient.models.observation import Observation as FHIR_Observation
+from fhirclient.models.fhirdate import FHIRDate
+
+class Patient(ABC):
+	class Gender(Enum):
+		male = 0
+		female = 1
+		other = 2
+		unknown = 3
+
+	def get_gender(self) -> Gender:
+		"""Returns the patient's gender. Default implementation return Gender.unkown"""		
+		return Gender.unknown
+
+	def get_identifier_value(self) -> str:
+		"""Returns a id value for external use."""
+		return None
+
+	@abstractmethod
+	def get_dob(self) -> datetime:
+		"""Returns the patient's date of birth."""
+		return datetime.min
+
+	def get_identifier_system(self) -> str:
+		"""Returns the namespace for unique identifier values."""
+		return None
+
+	def generate_name(self, gender : Gender) -> tuple[str, str]:
+		"""Generate a first and last name based on gender."""
+		match gender:
+			case Patient.Gender.male:
+				first_name = names.get_first_name('male')
+			case Patient.Gender.female:
+				first_name = names.get_first_name('female')
+			case default:
+				first_name = names.get_first_name()
+
+		return names.get_last_name(), first_name
+
+	def get_name(self) -> tuple[str, str]:
+		"""Returns a a tuple of the last and first name of the patient. Default implementation generates names based on gender"""
```

Comment:
Again, docstring needs to be clear that you're expected to return a tuple (last name, first name) in that exact order